app/src/format.py

```diff
@@ -86,5 +66,40 @@ def format_bem_documents(
     documents = _get_bem_documents_to_show(
         docs_shown_max_num, docs_shown_min_score, chunks_with_scores
     )
```

Comment:
Do we need `_get_bem_documents_to_show()` or `DocumentWithMaxScore`? Since we're showing all chunks, the code should be iterating through `chunks_with_scores` after it's been grouped by document.

Reply:
In case helpful, here's a function I wrote to group chunks by document, respecting max_num and min_score:

```python
def _get_bem_documents_to_show(
    docs_shown_max_num: int,
    docs_shown_min_score: float,
    chunks_with_scores: list[ChunkWithScore],
) -> OrderedDict[Document, list[ChunkWithScore]]:
    chunks_with_scores.sort(key=lambda c: c.score, reverse=True)

    # Build a dictionary of documents with their associated chunks,
    # Ordered by the highest score of each chunk associated with the document
    documents: OrderedDict[Document, list[ChunkWithScore]] = OrderedDict()
    for chunk_with_score in chunks_with_scores[:docs_shown_max_num]:
        document = chunk_with_score.chunk.document
        if chunk_with_score.score < docs_shown_min_score:
            logger.info(
                "Skipping chunk with score less than %f: %s",
                docs_shown_min_score,
                chunk_with_score.chunk.document.name,
            )
            continue

        if document in documents:
            documents[document].append(chunk_with_score)
        else:
            documents[document] = [chunk_with_score]

    return documents
    ```

Reply:
Here's an example of iterating over it to group the citations by document, for reference:

```python
    documents = _get_bem_documents_to_show(
        docs_shown_max_num, docs_shown_min_score, chunks_with_scores
    )

    html = "<h3>Source(s)</h3><ul>\n"

    for document in documents:
        html += f"<li>{_format_bem_title(document.name)}<ol>\n"

        for index, chunk in enumerate(documents[document], start=1):
            html += f"<li>Citation #{index} (score: {chunk.score})</li>\n"

        html += "</ol></li>\n"

    html += "</ul>"
```